[viu] Fixed bugs and API updated

zackmark29 opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Bug fix
- [X] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fix for: https://github.com/yt-dlp/yt-dlp/issues/3133

Playback stream API is now updated and  added method to get bearer token which required now for requesting playback streams

Also I just removed the old and unneeded function (_real_initialize)

pukkandan commented on yt_dlp/extractor/viu.py:
```diff
@@ -312,35 +295,50 @@ def _real_extract(self, url):
             'language_flag_id': self._LANGUAGE_FLAG.get(lang_code.lower()) or '3',
         }
+
+        rand = ''.join(random.choice('0123456789') for _ in range(10))
+        bearer_token = self._download_json(
+            f'https://api-gateway-global.viu.com/api/auth/token?v={rand}000',
+            video_id=None,
+            data=json.dumps({
+                'countryCode': country_code.upper(),
+                'platform': 'browser',
+                'platformFlagLabel': 'web',
+                'language': 'en',
+                'uuid': str(uuid.uuid4()),
+                'carrierId': '0'
+            }).encode('utf-8'),
+            headers={'Content-Type': 'application/json'},
+            note='Getting bearer token')['token']
```
put this in _real_initialize

zackmark29 replied:
> ~put this in _real_initialize~ cache the token in memory

Ah ok later after my lunch

zackmark29 replied:
I'm back. I'll update it now

zackmark29 replied:
Wait a minute I'm checking on something

zackmark29 replied:
@pukkandan Could you check?

zackmark29 replied:
@pukkandan can we just revert for now without caching the token?
I don't know yet how to parse the URL argument inside the _real__initialize to get the country code

pukkandan replied:
Which is why I edited my comment...
You can cache token in real_extract itself:

```py
auth_codes = {}
def _real_extract(self, url):
    ...
    country_code = ...
    if not self.auth_codes.get(country_code):
        self.auth_codes[country_code] = ...
    ...
```

zackmark29 replied:
> Which is why I edited my comment... You can cache token in real_extract itself:
> 
> ```python
> auth_codes = {}
> def _real_extract(self, url):
>     ...
>     country_code = ...
>     if not self.auth_codes.get(country_code):
>         self.auth_codes[country_code] = ...
>     ...
> ```

Thanks.

zackmark29 replied:
@pukkandan it's now updated. For MY and ID it's not possible for now. Their API is different and should have separated IE Class for that.

pukkandan commented on yt_dlp/extractor/viu.py:
```diff
@@ -227,21 +210,22 @@ class ViuOTTIE(InfoExtractor):
 
     def _raise_login_required(self):
         raise ExtractorError(
             'This video requires login. '
-            'Specify --username and --password or --netrc (machine: %s) '
-            'to provide account credentials.' % self._NETRC_MACHINE,
+            f'Specify --username and --password or --netrc (machine: {self._NETRC_MACHINE}) '
+            'to provide account credentials.',
             expected=True)
```
this function is actually redundant

zackmark29 replied:
Yeah. I only changed string format there.
But it's not required actually. It's just for old method which require token.

pukkandan commented on yt_dlp/extractor/viu.py:
```diff
@@ -312,35 +316,36 @@ def _real_extract(self, url):
             stream_data = None
             if video_data.get('user_level', 0) > 0:
-                user = self._login(country_code, video_id)
+                user = self._login(self._country_code, video_id)
```
mistake?
```suggestion
                user = self._login(country_code, video_id)
```

zackmark29 replied:
Yup I forgot to revert it. was going to change it if there's more to fix.

pukkandan commented:
> Also I just changed old style string format to f-string. Just let me know if that's not OK

Normally I'd rather not make unnecessary changes to extractors for easier mergeability with upstream. But ViuOTT is different enough from upstream that a refactor is justified

---

I think you broke `_login`. I have fixed it, and also refactored some code. Please confirm everything (including login) is working

zackmark29 commented:
> > Also I just changed old style string format to f-string. Just let me know if that's not OK
> 
> Normally I'd rather not make unnecessary changes to extractors for easier mergeability with upstream. But ViuOTT is different enough from upstream that a refactor is justified
> 
> I think you broke `_login`. I have fixed it, and also refactored some code. Please confirm everything (including login) is working

Yeah I didn't also touch other code only fixing the bug.

Ahm I didn't test and touch the login function. Since it's not needed anymore because of the bypass method for duration or Premium videos.

pukkandan commented:
> Ahm I didn't test and touch the login function. Since it's not needed anymore because of the bypass method for duration or Premium videos.

Would be nice if you could test it anyway (you can temporarily disable the bypass)

zackmark29 commented:
> > Ahm I didn't test and touch the login function. Since it's not needed anymore because of the bypass method for duration or Premium videos.
> 
> Would be nice if you could test it anyway (you can temporarily disable the bypass)

Alright

zackmark29 commented:
> > Ahm I didn't test and touch the login function. Since it's not needed anymore because of the bypass method for duration or Premium videos.
> 
> Would be nice if you could test it anyway (you can temporarily disable the bypass)

It's now updated. It takes long because of my connection =D
I can't do coding today because of work.
Feel free to clean it since I don't want to touch other codes,
But if there's any issue again, I'll check when I have time.

zackmark29 commented:
@pukkandan Do you think it's ok that we are doing bypass here for premium videos?
Or we'll just always required the user's to login to avoid any conflict in the future?

pukkandan commented:
> It's now updated. It takes long because of my connection =D
I can't do coding today because of work.
Feel free to clean it since I don't want to touch other codes,
But if there's any issue again, I'll check when I have time.

No worries. Take your time

pukkandan reviewed (approved):
Can be merged once someone tests the actual download

cc @BLSTR13, @uwuuuu

zackmark29 commented:
> Can be merged once someone tests the actual download
> 
> cc @BLSTR13, @uwuuuu

I guess @uwuuuu already tried to download. But I don't know if he got any error etc.
https://github.com/yt-dlp/yt-dlp/issues/3133#issuecomment-1074680988

BLSTR13 commented:
> > Can be merged once someone tests the actual download
> > cc @BLSTR13, @uwuuuu
> 
> I guess @uwuuuu already tried to download. But I don't know if he got any error etc. [#3133 (comment)](https://github.com/yt-dlp/yt-dlp/issues/3133#issuecomment-1074680988)

Already tested it, works fine

MrRawes commented:
just wondering @zackmark29 where the unlicense grant is, since every pull request in both youtube-dl and yt-dlp says check a box either grant under unlicense or provide valid evidense of the grant from someone else and [github itself has something similer](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license)

zackmark29 commented:
> just wondering @zackmark29 where the unlicense grant is, since every pull request in both youtube-dl and yt-dlp says check a box either grant under unlicense or provide valid evidense of the grant from someone else and [github itself has something similer](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license)

Oh I didn't notice about it. As I am not regularly doing this here. I thought it's referring to the whole previous code.
I looked on other PR and they checked "I am the original author of this code" even their code is only the fix. 
Should I do that then?

MrRawes commented:
depends if there were any changes from the original patch

------- Original Message -------

On Friday, March 25th, 2022 at 10:16 AM, zackmark29 ***@***.***> wrote:

> > just wondering @zackmark29 where the unlicense grant is, since every pull request in both youtube-dl and yt-dlp says check a box either grant under unlicense or provide valid evidense of the grant from someone else and github itself has something similer
> 

> Oh I didn't notice about it. As I am not regularly doing this here. I thought it's referring to the whole previous code.
> 

> I looked on other PR and they checked "I am the original author of this code" even their code is only the fix.
> 

> Should I do that then?
> 

> —
> 

> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

> You are receiving this because you commented.Message ID: ***@***.***>
